cmd/manager/main.go

```diff
@@ -183,8 +182,14 @@ func main() {
 		os.Exit(1)
 	}
 
-	// Add the Metrics Service
-	addMetrics(ctx, cfg)
+	err = mgr.Add(manager.RunnableFunc(func(stop <-chan struct{}) error {
+		createMetricsResources(context.Background(), cfg, mgr.GetClient(), dc)
+		return nil
+	}))
```

Comment:
Can you please tell me what the purpose of `manager.RunnableFunc` is?

I read [its docs](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Runnable) but I don't understand why this is used here instead of calling `createMetricsResources` directly. Also, when is `Start` called on the `RunnableFunction`?

Reply:
Oh, sorry, I didn't see this in the description. Makes sense.

Reply:
If we call `createMetricsResources()` directly (instead of having the manager call it), the function will fail when it tries to read from the manager's client. This happens because the client's caches aren't synced until the manager starts. The only reason we could previously call the function directly was because it was creating its own client.

In contrast, when the function is called via a `RunnableFunc` registered via `mgr.Add()`, the manager waits for the [caches to be synced before invoking the registered runnables](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.6.3/pkg/manager/internal.go#L579-L586). 

So, if you want to use the manager's client, you have to do it in a runnable registered with the manager.

Reply:
Thank you for explaining! Great to know about this.